### PR TITLE
Add update mechanism for :penguin:

### PR DESCRIPTION
**Pull request by stanier**
Temporary(?) solution to #2956 while better solution is discussed.

### CODE REVIEW

**Comment by p-e-w** [comment]
This PR is a good shot at fixing the single biggest issue on Linux. It has now been tagged "needs-review" for almost a month. Any chance of speeding up the review process so there can be progress on this issue?

The way I see it, while the approach of trying to identify the distro and then downloading packages is not pretty, it's the only way to move forward given the circumstances. There is one – and only one – proper way to install and update packages on Linux: From repositories. If the Atom team is unwilling to create official aptitude/yum repos, the "semi-manual" approach is the next best thing. Using third-party repos is simply not an option in many environments due to security concerns.

**Comment by antond** [comment]
This would be great to have. I hope it gets merged soon.

**Comment by Zireael07** [comment]
Has it been reviewed at all? There are no comments to this effect...

**Comment by stanier** [comment]
@Zireael07 I can't really attest to the functionality of this fix as it was only tested on two local builds and I didn't really care to mess with CoffeeScript's syntactic bull.  I think the only part that really needs "reviewing" is the decision of whether or not to continue using the release channel for updates or to make use of a ppa (as we await a response in #2956).

In my personal opinion, the update process should be automated regardless.  I might push a commit later today to touch up some things and make it more likely to be implemented.

File: src/browser/auto-update-manager.coffee [hunk 1]
```diff
@@ -41,8 +43,8 @@ class AutoUpdateManager
 
     autoUpdater.on 'checking-for-update', =>
       @setState(CheckingState)
-
     autoUpdater.on 'update-not-available', =>
+
```
Comment (50Wliu) [inline src/browser/auto-update-manager.coffee]: Why was this line added, and why was the above one removed?

File: src/browser/auto-update-manager.coffee [hunk 2]
```diff
@@ -59,7 +61,7 @@ class AutoUpdateManager
       when 'win32'
```
Comment (50Wliu) [inline src/browser/auto-update-manager.coffee]: This can be combined into `when 'win32', 'linux'`

File: src/browser/auto-updater-linux.coffee [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    error?.code ?= code
+    error?.stdout ?= stdout
+    callback?(error, stdout)
+
+class AutoUpdater
+  _.extend @prototype, EventEmitter.prototype
+
+  stdout = ChildProcess.execSync('cat /etc/issue').toString().replace(/\\./g, "")
+
+  if distros.ubuntu.test(stdout) | distros.debian.test(stdout)
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: Shouldn't this be using `||` instead of `|`?

**Comment by stanier** [comment]
Passes unit tests now.

As a side note, this is why I generally dislike CoffeeScript (see a6d4b91 d9c68a4)

File: src/browser/auto-updater-linux.coffee [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        return
+
+      @emit 'update-available'
+
+      @installUpdate (error) =>
+        if error?
+          @emit 'update-not-available'
+          return
+
+        console.log 'will update'
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: ;o

File: src/browser/auto-updater-linux.coffee [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  spawnedProcess.on 'close', (code, signal) ->
+    error ?= new Error("Command failed: #{signal ? code}") if code isnt 0
+    error?.code ?= code
+    error?.stdout ?= stdout
+    callback?(error, stdout)
+
+class AutoUpdater
+  _.extend @prototype, EventEmitter.prototype
+
+  stdout = ChildProcess.execSync('cat /etc/issue').toString().replace(/\\./g, "")
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: I believe Atom uses `lsb_release` to determine the OS - could you do that instead?
https://github.com/atom/notifications/blob/4716ca24f74c685f6db85447e432e8a173d86a67/lib/user-utilities.coffee#L64-L78 (though it does seem a bit broken on Fedora)

File: src/browser/auto-updater-linux.coffee [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    @pkgType = 'deb'
+    @installCmd = 'dpkg --install -y'
+  else if distros.redhat.test(stdout) or distros.centos.test(stdout)
+    @pkg = 'atom.x86_64.rpm'
+    @pkgType = 'rpm'
+    @installCmd = 'yum localinstall -y'
+  else if distros.fedora.test(stdout)
+    @pkg = 'atom.x86_64.rpm'
+    @pkgType = 'rpm'
+    @installCmd = 'dnf install -y'
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: Does this work on Fedora 21 and under?

File: src/browser/auto-updater-linux.coffee [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    app.quit()
+
+  downloadUpdate: (callback) ->
+    spawn 'curl', ['-L', @updateUrl], (error, stdout) ->
+      return callback(error) if error?
+
+      try
+        update = JSON.parse(stdout)
+
+        spawn 'atom', ['--version'], (error, stdout) ->
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: The command isn't necessarily `atom` anymore with the introduction of beta releases - `atom-beta` is valid now and will launch the beta version which also supports auto-updates.  You'll need to catch that.

File: src/browser/auto-updater-linux.coffee [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+  setFeedUrl: (@updateUrl) ->
+
+  quitAndInstall: ->
+    app.once 'will-quit', -> spawn 'sudo', [installCmd, @pkg, '&&',
+      'rm', '-rf', @pkg, '&&', 'atom']
+    app.quit()
+
+  downloadUpdate: (callback) ->
+    spawn 'curl', ['-L', @updateUrl], (error, stdout) ->
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: Is there any chance at all that `curl` isn't installed?  If so we'll need to add it as a dependency.

File: src/browser/auto-updater-linux.coffee [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  else if distros.fedora.test(stdout)
+    @pkg = 'atom.x86_64.rpm'
+    @pkgType = 'rpm'
+    @installCmd = 'dnf install -y'
+  else distro = 'unsupported'
+
+  setFeedUrl: (@updateUrl) ->
+
+  quitAndInstall: ->
+    app.once 'will-quit', -> spawn 'sudo', [installCmd, @pkg, '&&',
```
Comment (50Wliu) [inline src/browser/auto-updater-linux.coffee]: `sudo` :(
I'm guessing that this can't be avoided though.

**Comment by 50Wliu** [comment]
The linux packages are only for 64-bit right now, but I'm fairly certain that 32-bit users can build it manually and run Atom (possibly pinning it to a release version and enabling updates).  If so, that should be addressed in the auto-updater so that it doesn't try installing 64-bit builds on 32-bit machines.

**Comment by stanier** [comment]
... I'll just wait for the official repositories

**Comment by lee-dohm** [comment]
Thanks @stanier for all your hard work on this. We talked this over and we're not comfortable with supporting a partial solution like this. We're still planning on having support for the official repositories. I don't have an ETA on when that will be accomplished right now though.

Thanks again for driving this!
